### PR TITLE
[WIP] Implement API for branches listing

### PR DESCRIPTION
**Pull request by sapk**
Implement feature request : #2410

Need https://github.com/gogits/go-gogs-client/pull/15

### CODE REVIEW

**Comment by bkcsoft** [comment]
Looks nice, though I see a lot of TODO-comments that relate to gogits/git-module, are those supposed to be there still or is it something that just came along with the commit? :smile:

Also for commits that you know won't build (because depending PRs havn't been merged yet) you could add `[ci skip]` to the commit-message (preferably at a new-line so it doesn't show that well when scrolling through the commit-history) :smile: see https://docs.travis-ci.com/user/customizing-the-build/#Skipping-a-build for more info

**Comment by sapk** [comment]
There some TODO related to update the wiki for API calls. I will do it soon and i need to review the code. I PR to discuss if i done it right and have some advices. Sorry i didn't precise. I will polish the code.

**Comment by sapk** [comment]
Related Wiki page (draft): https://gist.github.com/sapk/df64347ff218baf4a277
the url link is not implemented since there is no API for commit like https://developer.github.com/v3/repos/commits/

File: models/branch.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package models
```
Comment (Unknwon) [inline models/branch.go]: I think the file name should be renamed to `repo_branch.go`.

File: models/repo.go [hunk 2]
```diff
@@ -288,6 +288,20 @@ func (repo *Repository) GetMirror() (err error) {
...
+	if(!git.IsBranchExist(repo.RepoPath(), br)){
+		return nil, errors.New("Branch do not exist");
+	}
+	return &Branch{
+			Path: repo.RepoPath(),
+			Name: br,
+	},nil
+}
+
+func (repo *Repository) GetBranches() (_ []*Branch, err error) {
```
Comment (Unknwon) [inline models/repo.go]: You can just use `func (repo *Repository) GetBranches() ([]*Branch, error) {`

File: models/branch.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	for i := range brs {
+		Branches[i] = &Branch{
+			Path: path,
+			Name: brs[i],
+		}
+	}
+	return Branches, nil
+}
+
+func GetBranchesByRepo(user,repo string) ([]*Branch, error) {
```
Comment (Unknwon) [inline models/branch.go]: Seems not been used anywhere?

File: models/repo.go [hunk 4]
```diff
@@ -288,6 +288,20 @@ func (repo *Repository) GetMirror() (err error) {
 	return err
 }
 
+func (repo *Repository) GetBranch(br string) (_ *Branch, err error) {
+	if(!git.IsBranchExist(repo.RepoPath(), br)){
+		return nil, errors.New("Branch do not exist");
```
Comment (Unknwon) [inline models/repo.go]: Should use `fmt.Errorf("Branch does not exist: %s", br)`.

File: models/repo.go [hunk 5]
```diff
@@ -288,6 +288,20 @@ func (repo *Repository) GetMirror() (err error) {
 	return err
 }
 
+func (repo *Repository) GetBranch(br string) (_ *Branch, err error) {
```
Comment (Unknwon) [inline models/repo.go]: Should use `(*Branch, error)`.

File: routers/api/v1/convert/convert.go [hunk 6]
```diff
@@ -10,11 +10,32 @@ import (
 	"github.com/Unknwon/com"
 
 	api "github.com/gogits/go-gogs-client"
+	"github.com/gogits/git-module"
 
 	"github.com/gogits/gogs/models"
 	"github.com/gogits/gogs/modules/setting"
 )
 
+// ToApiBranch converts user to its API format.
```
Comment (Unknwon) [inline routers/api/v1/convert/convert.go]: This part of code should put after at least later then `ToApiRepository` as `branch` belongs to a `repository`.

File: routers/api/v1/repo/branch.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
+// Copyright 2014 The Gogs Authors. All rights reserved.
```
Comment (Unknwon) [inline routers/api/v1/repo/branch.go]: It's `2016`.

Also add this copyright content to `models/repo_branch.go` file.

File: routers/api/v1/api.go [hunk 8]
```diff
@@ -183,7 +183,10 @@ func RegisterRoutes(m *macaron.Macaron) {
 				m.Patch("/hooks/:id:int", bind(api.EditHookOption{}), repo.EditHook)
 				m.Get("/raw/*", middleware.RepoRef(), repo.GetRawFile)
 				m.Get("/archive/*", repo.GetArchive)
-
+				m.Group("/branches", func() {
+					m.Get("",repo.ListBranches)
+					m.Get("/:id",repo.GetBranch)
```
Comment (Unknwon) [inline routers/api/v1/api.go]: Shouldn't we use `:branchname` here?

File: routers/api/v1/repo/branch.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	api "github.com/gogits/go-gogs-client"
+
+	"github.com/gogits/gogs/modules/middleware"
+	"github.com/gogits/gogs/routers/api/v1/convert"
+)
+
+// Temporary: https://gist.github.com/sapk/df64347ff218baf4a277#get-a-branch
+// https://github.com/gogits/go-gogs-client/wiki/Repositories-Branches#get-a-branch
+func GetBranch(ctx *middleware.Context) {
+	//Getting the branch requested
```
Comment (Unknwon) [inline routers/api/v1/repo/branch.go]: Please add an empty space after `///`.

**Comment by Unknwon** [comment]
Thanks your PR!

Comments are made to improve it.

File: models/branch.go (hunk 3 above)
Comment (sapk) [inline models/branch.go]: That an old artefact from a first try but after i have found that ctx.Repo.Repository is prefill under api so i use a direct call ctx.Repo.Repository.GetBranches(). Should i comment it or remove it completely ?

Comment (Unknwon) [inline models/branch.go]: I think you can remove it for now.

**Comment by sapk** [comment]
Thanks for the comments. I rebase code on develop and found that i choose the wrong branch to PR.
I will re-open a PR on develop. Sorry ^^

**Comment by sapk** [comment]
Re-open for PR on develop at #2506. Closing this one.

**Pull request closed without merging**